File: All Nodes Distance K in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        if k == 0:
            return [target.val]

        ans = []

        def isTargetInTree(node):
            nonlocal ans

            if node is None:
                return False, 0

            if node is target:
                findNodeByDist(node, k)
                return True, 1

            inLeftTree, leftDist = isTargetInTree(node.left)
            inRightTree, rightDist = isTargetInTree(node.right)
            dist = 0

            if inLeftTree:
                dist = leftDist
                if leftDist == k:
                    ans.append(node.val)
                elif k > leftDist:
                    findNodeByDist(node.right, k - leftDist - 1)

            if inRightTree:
                dist = rightDist
                if rightDist == k:
                    ans.append(node.val)
                elif k > rightDist:
                    findNodeByDist(node.left, k - rightDist - 1)
            return (inLeftTree or inRightTree), dist + 1

        def findNodeByDist(node, dist):
            nonlocal ans
            if node is None:
                return

            if dist == 0:
                ans.append(node.val)

            findNodeByDist(node.left, dist - 1)
            findNodeByDist(node.right, dist - 1)

        isTargetInTree(root)
        return ans

File: All Nodes Distance K in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    unordered_map<int, TreeNode*> parents;

    void addParent(TreeNode* node, TreeNode* parent) {
        if (!node) return;

        parents[node->val] = parent;
        addParent(node->left, node);
        addParent(node->right, node);
    }

    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        addParent(root, nullptr);

        vector<int> ans;
        deque<pair<int, TreeNode*>> q;
        unordered_set<int> visited;

        q.emplace_back(0, target);

        while (q.size() > 0) {
            auto [dist, node] = q.front();
            q.pop_front();
            visited.insert(node->val);

            if (dist == k) {
                ans.push_back(node->val);
            } else {
                if (node->right && !visited.contains(node->right->val)) q.emplace_back(dist+1, node->right);
                if (node->left && !visited.contains(node->left->val)) q.emplace_back(dist+1, node->left);
                if (parents[node->val] && !visited.contains(parents[node->val]->val)) q.emplace_back(dist+1, parents[node->val]);
            }
        }

        return ans;
    }
};
